nova/volume/cinder.py

```diff
@@ -561,28 +561,36 @@ def detach(self, context, volume_id, instance_uuid=None,
         client = cinderclient(context)
         if attachment_id is None:
             volume = self.get(context, volume_id)
+            attachments = volume.get('attachments', {})
+            if instance_uuid:
+                attachment_id = attachments.get(instance_uuid, {}).\
+                        get('attachment_id')
+            else:
+                LOG.warning(_LW("attachment_id couldn't be retrieved for "
+                                "volume %(volume_id)s. The volume has the "
+                                "'multiattach' flag enabled, without the "
```

Comment:
At this point we don't know if the volume is multiattach or not. Perhaps this log should be moved back under the multiattach `if`.